### PR TITLE
Don't create an XLF file if there is nothing to translate

### LINKED ISSUE #64: Allow opting out individual files, or don't produce xlf file if no translatable elements exist
I have a XamlRule file in a project where Xlf translations are turned on (and desired) but when an individual file has no translatable attributes (eg: https://github.com/davidwengier/project-system/blob/SplitFileAndBrowseObjectNodes/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/None.xaml) we still get a dozen or so basically empty xlf files generated

https://github.com/davidwengier/project-system/blob/SplitFileAndBrowseObjectNodes/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Rules/xlf/None.xaml.fr.xlf

It would be ideal to prevent this in some way, presumably either via some metadata that can opt out a file in the project file, or simply by not creating an xlf file, or attempting to update same, until there is something to write.

### PR DESCRIPTION
**Pull request by davidwengier**
I'm not 100% sure if this was the right spot for the change so let me know if I'm way off base. I haven't tested this in Visual Studio, only via the tests I added. Would love to know what the next steps are.

Fixes #64

### CODE REVIEW

**Comment by nguerrera** [comment]
I think I am wrong. My own comment explains why xlf file timestamps can't be used for incrementality :)

https://github.com/dotnet/xliff-tasks/blob/977ba73e848b0c01e3c265cbba37aa9807f2eba7/src/XliffTasks/build/XliffTasks.targets#L21-L29

File: src/XliffTasks.Tests/UpdateXlfTests.cs [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+                var task = new UpdateXlf();
+                task.AllowModification = true;
+                task.Languages = new string[] { "fr" };
+                task.TranslateDocument(sourceFile, "XamlRule", xlfPath);
+
+                Assert.True(File.Exists(Path.Combine(xlfPath, "xlf.fr.xlf")));
+            }
+            finally
+            {
+                Directory.Delete(tempPath, true);
```
Comment (nguerrera) [inline src/XliffTasks.Tests/UpdateXlfTests.cs]: Am I reading wrong or are both tests using the same tempPath? Won't this break when tests are run in parallel?

**Comment by nguerrera** [comment]
Hmm, ok, now I think there is a problem again. :)

There is a place where we use the xlf files as *inputs* to avoid regenerating the translated files on every build: If the xlf hasn't changed and the source hasn't changed, then we don't need to generate the translated file:

https://github.com/dotnet/xliff-tasks/blob/977ba73e848b0c01e3c265cbba37aa9807f2eba7/src/XliffTasks/build/XliffTasks.targets#L164-L165

If an input does not exist, then msbuild will run the target:


``` xml
<Project>
  <Target Name="X" Inputs="NotAFileThatExists.txt" Outputs="Output.txt">
    <Touch Files="Output.txt" AlwaysCreate="true" />
  </Target>
</Project>
```

msbuild /v:detailed more than once
```
Input file "NotAFileThatExists.txt" does not exist.
Using "Touch" task from assembly "Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a".
Task "Touch"
  Touching "Output.txt".
```

**Comment by nguerrera** [comment]
I think it will be difficult to get the incremental behavior right if the XLF file doesn't get created. It adds adds more incremental cases: first string added, last string removed.

How cumbersome would it be for you to use XlfInput="false" metadata for files that don't have translations? Perhaps organizing the xaml files in way that you can set this metadata via a glob?

 We should weigh that against adding complexity here.

**Comment by nguerrera** [comment]
Sigh, for the project system rules, it won't work to turn of translation altogether with XlfInput=false. 

Because every xaml file has to be in every localized directory since there isn't a probe per xaml file, but rather one big probe  for the xaml directory here:

https://github.com/dotnet/project-system/blob/615032e1c515df5e2c064d3b803932ee7aeaa09e/src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/DesignTimeTargets/Microsoft.Managed.DesignTime.targets#L27

**Comment by nguerrera** [comment]
Actually, I guess there's nothing stopping you from updating those targets appropriately. If you had some convention for what is localizable and what isn't, then you could handle that both during the build to set XlfInput=false and during design-time evaluation to point to the right directory. This would have the advantage of removing unnecessary duplicated files that are never translated.

(Sorry for going back and forth, basically thinking out loud here.)

**Comment by davidwengier** [comment]
Thank you! I rather appreciate the thinking out loud, helps to have context as I'm coming up to speed on this stuff. I think `XlfInput` will work fine for my needs, and there is a convention I should be able to use. I'll have to figure out why I didnt discover that option when I went looking.

If that works then I will happily close this PR and close the issue.

File: src/XliffTasks.Tests/XliffTasks.Tests.csproj [hunk 2]
```diff
@@ -6,8 +6,18 @@
     <TargetFramework>netcoreapp2.0</TargetFramework>
   </PropertyGroup>
 
+  <ItemGroup>
+    <Compile Remove="obj\**" />
+    <EmbeddedResource Remove="obj\**" />
+    <None Remove="obj\**" />
+  </ItemGroup>
```
[the remark this answers is not in the document]
Comment (davidwengier) [reply src/XliffTasks.Tests/XliffTasks.Tests.csproj]: Build failures locally if the obj folder was included. Seemed to me like it shouldn't have existed but I had no local modifications

**Comment by nguerrera** [comment]
>  I'll have to figure out why I didnt discover that option when I went looking.

That;s one me. I thought it would be useful at some point when I coded it,  but I didn't document it. Also,  other than being wired to non-true for non resx EmbeddedResource items, it hasn't had any usage. So it's possible that it doesn't work as I intended right now. So let me know how it goes. :)

**Comment by davidwengier** [comment]
XlfInput works great

**Pull request closed without merging**